### ISSUE TITLE
"not pinned or pinned indirectly" - request for more specifics

### ISSUE THREAD

**Issue by obo20** [ipfs/go-ipfs#7440]
When unpinning content from IPFS, it's possible to get the error message:

"not pinned or pinned indirectly"

Would it be possible to have a specific error message for each of these two scenarios? 

In practice, these two scenarios are very different. For us, one scenario means that something should have been pinned when in reality it somehow wasn't, which is very bad. The other scenario means that an item is simply referenced by a different pin in our system, which isn't bad at all.

**Comment by Stebalien** [comment]
How about "not directly pinned (but may be indirectly pinned)"? We _used_ to actually _check_ to see if the data was indirectly pinned, but that meant `ipfs pin ls QmFoobar` would take _ages_ to return an error for unpinned content because it had to traverse all pinned content to determine if it was indirectly pinned.

**Comment by obo20** [comment]
@Stebalien The message isn't so much the issue as it is the inability to determine what type of error you're receiving. 

From what you're saying though it sounds like making this distinction isn't as easy as it sounds. (I was hoping this might be an easy fix) 

It sounds like this might be something to revisit in the future when a reference counting type system comes into play.

**Comment by Stebalien** [comment]
Yeah, that's basically it. And yes, once we get reference counting, this should be easier. But even then, that depends on _how_ we do reference counting (some optimizations may involve lazily updating reference counts on demand).

**Comment by gammazero** [comment]
@obo20 In the cases when you encounter the "not pinned or pinned indirectly" error message, you can check if the object is pinned indirectly using `ipfs pin ls -t indirect <cid>`.  If the object is pinned indirectly, this will tell you so and tell what recursive pin it is pinned by.  This may take some time, since this means searching potentially every CID that every recursive pin has pinned.  So. hopefully that is the rare exception for you.

The PR linked above offers this suggestion in the help details for the "pin rm" command.
